Std/Tactic/Pattern/Utils.lean

```diff
@@ -0,0 +1,163 @@
+    return .all
+  match stx[0] with
+  | `(occs| (occs := *)) => return .all
+  | `(occs| (occs := $ids*)) =>
+    return .pos <| Array.toList <| ← ids.mapM fun id =>
+      let n := id.toNat
+      if n == 0 then
+        throwErrorAt id "positive integer expected"
+      else return n
+  | _ => throwError m! "{stx}"
+
+/-- Elaborate the occurrences, pattern and location in a  `patternLocation`. -/
+def expandPatternLocation (stx : Syntax) : TermElabM (Occurrences × AbstractMVarsResult × Location) := do
+  let occs ← expandOptOccs stx[0]
+  let pattern ← expandPattern stx[1]
+  let loc := expandOptLocation stx[2]
+  return (occs, pattern, loc)
+section Expand
+
+
+
+end Expand
+
+section PatternsAndOccurrences
+
+/-- The pattern at a given position in an expression.
+    Variables under binders are turned into meta-variables in the pattern. -/
+def SubExpr.patternAt (p : SubExpr.Pos) (root : Expr) : MetaM Expr := do
+  let e ← Core.viewSubexpr p root
+  let binders ← Core.viewBinders p root
+  let mvars ← binders.mapM fun (name, type) =>
+    mkFreshExprMVar type (userName := name)
+  return e.instantiateRev mvars
+
+/-- Finds the occurrence number of the pattern in the expression
+    that matches the sub-expression at the specified position.
+    This follows the code of `kabstract` from Lean core. -/
+def findMatchingOccurrence (position : SubExpr.Pos) (root : Expr) (pattern : Expr) : MetaM Nat := do
+  let root ← instantiateMVars root
+  unless ← isDefEq pattern (← SubExpr.patternAt position root) do
+    throwError s!"The specified pattern does not match the pattern at position {position}."
+  let pattern ← instantiateMVars pattern
+  let pHeadIdx := pattern.toHeadIndex
+  let pNumArgs := pattern.headNumArgs
+  let rec
+  /-- The recursive step in the expression traversal to search for matching occurrences. -/
+  visit (e : Expr) (p : SubExpr.Pos) (offset : Nat) := do
+    let visitChildren : Unit → StateRefT Nat (OptionT MetaM) Unit := fun _ => do
+      match e with
+      | .app f a         => do
+        visit f p.pushAppFn offset <|>
+        visit a p.pushAppArg offset
+      | .mdata _ b       => visit b p offset
+      | .proj _ _ b      => visit b p.pushProj offset
+      | .letE _ t v b _  => do
+        visit t p.pushLetVarType offset <|>
+        visit v p.pushLetValue offset <|>
+        visit b p.pushLetBody (offset+1)
+      | .lam _ d b _     => do
+        visit d p.pushBindingDomain offset <|>
+        visit b p.pushBindingBody (offset+1)
+      | .forallE _ d b _ => do
+        visit d p.pushBindingDomain offset <|>
+        visit b p.pushBindingBody (offset+1)
+      | _                => failure
+    if e.hasLooseBVars then
+      visitChildren ()
+    else if e.toHeadIndex != pHeadIdx || e.headNumArgs != pNumArgs then
+      visitChildren ()
+    else if (← isDefEq e pattern) then
+      let i ← get
+      set (i+1)
+      if p = position then
+        return ()
+      else
+        visitChildren ()
```

Comment:
This code in `kabstract` has been updated in the meantime to give better behaviour!